datafusion/common/src/scalar.rs

```diff
@@ -739,12 +759,21 @@ macro_rules! build_timestamp_array_from_option {
 
 macro_rules! eq_array_primitive {
     ($array:expr, $index:expr, $ARRAYTYPE:ident, $VALUE:expr) => {{
-        let array = $array.as_any().downcast_ref::<$ARRAYTYPE>().unwrap();
```

Comment:
This error can be unreachable if an invariant is established that the `array` passed to [`eq_array`](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L1664) always has the same data type, precision and scale with `self`.
I will change this back to panic and thus making `eq_array` panic instead of returning Result.